core/arch/arm/kernel/user_ta.c

```diff
@@ -613,75 +470,18 @@ static TEE_Result user_ta_enter(TEE_ErrorOrigin *err,
 	return res;
 }
 
-/*
- * Load a TA via RPC with UUID defined by input param uuid. The virtual
- * address of the TA is recieved in out parameter ta
- *
- * Function is not thread safe
- */
-static TEE_Result rpc_load(const TEE_UUID *uuid, struct shdr **ta,
-			uint64_t *cookie_ta)
-{
-	TEE_Result res;
-	struct optee_msg_param params[2];
-	paddr_t phta = 0;
-	uint64_t cta = 0;
-
-
-	if (!uuid || !ta || !cookie_ta)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	memset(params, 0, sizeof(params));
-	params[0].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
-	tee_uuid_to_octets((void *)&params[0].u.value, uuid);
-	params[1].attr = OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT;
-	params[1].u.tmem.buf_ptr = 0;
-	params[1].u.tmem.size = 0;
-	params[1].u.tmem.shm_ref = 0;
-
-	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_LOAD_TA, 2, params);
-	if (res != TEE_SUCCESS)
-		return res;
-
-	thread_rpc_alloc_payload(params[1].u.tmem.size, &phta, &cta);
-	if (!phta)
-		return TEE_ERROR_OUT_OF_MEMORY;
-
-	*ta = phys_to_virt(phta, MEM_AREA_NSEC_SHM);
-	if (!*ta) {
-		res = TEE_ERROR_GENERIC;
-		goto out;
-	}
-	*cookie_ta = cta;
-
-	params[0].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
-	tee_uuid_to_octets((void *)&params[0].u.value, uuid);
-	params[1].attr = OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT;
-	params[1].u.tmem.buf_ptr = phta;
-	params[1].u.tmem.shm_ref = cta;
-	/* Note that params[1].u.tmem.size is already assigned */
-
-	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_LOAD_TA, 2, params);
-out:
-	if (res != TEE_SUCCESS)
-		thread_rpc_free_payload(cta);
-	return res;
-}
-
-static TEE_Result init_session_with_signed_ta(const TEE_UUID *uuid,
-				const struct shdr *signed_ta,
+static TEE_Result init_session_with_ta_store(const TEE_UUID *uuid,
+				const struct user_ta_store_ops *ta_store,
 				struct tee_ta_session *s)
 {
 	TEE_Result res;
 
-	DMSG("   Load dynamic TA");
-	/* load and verify */
-	res = ta_load(uuid, signed_ta, &s->ctx);
+	DMSG("Load user TA %pUl", (void *)uuid);
+	res = ta_load(uuid, ta_store, &s->ctx);
 	if (res != TEE_SUCCESS)
 		return res;
-
-	DMSG("      dyn TA : %pUl", (void *)&s->ctx->uuid);
-
+	if (memcmp(uuid, &s->ctx->uuid, sizeof(TEE_UUID)));
+		panic();
```

Comment:
This "can't happen" since `ta_load()` already does a `memcmp()` and returns an error on mismatch.  The `assert()` should be good enough if that's even needed.